docs/source/theory/supportedEquationSet.rst

```diff
@@ -422,6 +402,104 @@ following temperature form:
+:math:`\bar{\rho} = \left< \bar{\rho} \right> + \bar{\rho}'`.
+A decomposition of the plane averaged momentum at a given instance in time is then
+
+.. math::
+       \left< \bar{\rho}  \mathrm{u}_i  \right>  =
+        \left< \bar{\rho} \right> \left< \mathrm{u}_i \right>
+        + \left< \bar{\rho}'  \mathrm{u}'_i  \right>.
+
+We now wish to apply a momentum source based on a desired spatial averaged velocity
+:math:`\mathrm{U}_i`.
+This can be expressed as:
+
+.. math::
+       \left< \bar{\rho}  \mathrm{u}_i^*  \right>  =
+        \left< \bar{\rho} \right> \left< \mathrm{u}^*_i \right>
+        + \left< \bar{\rho}'  {\mathrm{u}^*_i}'  \right>,
+
+where :math:`\mathrm{u}_i^*` is an unknown reference velocity field whose volume
+average is the desired  velocity :math:`\left< \mathrm{u}_i^* \right> = \mathrm{U}_i`.
+Since the correlation :math:`\left< \bar{\rho}'  \mathrm{u^*}'_i  \right>`
+is unknown, we assume that
+
+.. math::
+    \left< \bar{\rho}'  \mathrm{u^*}'_i  \right>
+    =
+    \left< \bar{\rho}'  \mathrm{u}'_i  \right>
+
+such that the momentum source can now be defined as:
+
+.. math::
+   :label: abl-mom-source
+
+   {\mathrm{f}_i} = \alpha_u
+        \left(  \, \frac{\left< \bar{\rho} \right> \mathrm{U_i}
+        - \left< \bar{\rho} \right> \left< \mathrm{u}_i \right>}
+        {\Delta t}\right)
+
+where :math:`\left< \right>` denotes volume averaging at a
+certain time :math:`t`,
+:math:`\mathrm{U}_i` is the desired spatial averaged
+velocity,
+and :math:`\Delta t` is the time-scale between when the source term is computed
+(time :math:`t`) and when it is applied (time :math:`t + \Delta t`).
+This is typically chosen to be the simulation time-step.
+In the case of an ABL simulation with flat terrain, the voulme averaging is done
+over an infinitesimal small plane over the :math:`x` and :math:`y` directions,
+such that the body force is only a
+function of height :math:`z` and time :math:`t`.
+The implementation allows the
+user to prescribe relaxation factors :math:`\alpha_u` for the source terms that are
+applied. Nalu uses a default value of 1.0 for the relaxation factors if no
+values are defined in the input file during initialization.
+
+The enthalpy source term works similarly to the momentum source term.
+A temperature difference is computed at every time-step and a forcing term
+is added to the enthalpy equation:
+
+.. math::
+
+  S_\theta = \alpha_\theta C_p
+      \left(
+         \frac{\theta_{\rm ref} - \left< \theta \right>}{\Delta t}
+      \right)
+
+where :math:`\theta_{\rm ref}` is the desired spatial averaged temperature,
+:math:`\left< \theta \right>` is the spatial averaged temperature,
+:math:`C_p` is the heat capcity,
+:math:`\alpha_\theta` is the relaxation factor,
+and
+:math:`\Delta t` is the time-scale.
+
+The present implementation can vary the
+source terms as a function of time and space using either a user-defined table
+of previously computed source terms (e.g., from a *precursor* simulation or
```

Comment:
Is the user_defined forcing implemented?

Reply:
@mbarone81 Yes, it is implemented, but hasn't seen a lot of testing though. We will have to test it when we get to the inflow/outflow simulations.

Reply:
Can you point me to it for my own sake?  I missed it....

Reply:
@mbarone81 take a look at the method `compute_momentum_sources` https://github.com/tonyinme/Nalu/blob/1d97ae17a093dc53aff0cf3616992f9954ac6dc1/src/ABLForcingAlgorithm.C#L451. The `else` block implements `user_defined` in this case the `velocity_x`, `velocity_y` entries in input file are interpreted as source terms instead of a target velocity.

Reply:
I guess some optimizations could be implemented to skip transfers when the forcing is `user_defined` currently it will do the search without consideration of that flag.

Reply:
Good point.  It would also be good to alleviate the requirement to have a from_target_part specified when the forcing is user_defined.

Reply:
Agreed, in general we could minimize the places where we need target parts... by default we should be able to pull interior/material_property parts and use those.

Reply:
@tonyinme Do you think that modification to skip transfers when the forcing is user_defined is a straightforward task that fits with the resources of the related story?  If so, I recommend doing it for this pull request.  If it is a bigger effort then I recommend that it be captured in a subsequent story.

Reply:
Hey @mbarone81, I am not sure, it does not look straight forward to me. @sayerhs, would this be easy to do?

Reply:
@tonyinme My first guess is it should be straightforward with a few `if` conditions, can we sit together tomorrow and look at the code and decide? We can report back at tomorrow's standup. @mbarone81 will that be ok?

Reply:
Yes, sure, that would be fine.

Reply:
sounds good to me!